### PR TITLE
Source Recurly: Add dbt converter  

### PR DESCRIPTION
**Pull request by btkcodedev**
Closes https://github.com/airbytehq/airbyte-dbt-models/issues/24

**How**
1) The tables from the fivetran model are analyzed for source-recurly
https://github.com/fivetran/dbt_recurly_source/tree/main/models

2) Fivertran's Streams were matched against Airbyte streams and differential tables were identified
3) Single transformation for the table `account_balance` is made from the account stream

Raising 1st draft PR for review

| model | status |
| ---- | --- |
| account_balance_history | 🟩  |
| account_history | 🟩 |
...

### CODE REVIEW

**Comment by btkcodedev** [comment]
Current analysis: @marcosmarxm 

URL: https://fivetran.github.io/dbt_recurly/#!/model/model.recurly_source

| **Fivetran**                                    | **Airbyte**            |
|-------------------------------------------------|------------------------|
| `stg_recurly__account_balance_history.sql`      | Trasformation from `accounts`  |
| `stg_recurly__account_history.sql`              | `accounts`  |
| `stg_recurly__account_note_history.sql`         | `account_notes`        |
| `stg_recurly__billing_info_history.sql`         | `billing_info`         |
| `stg_recurly__coupon_discount.sql`              | `coupons`              |
| `stg_recurly__coupon_redemption_history.sql`    | ❌ it doesn't exist         |
| `stg_recurly__credit_payment_history.sql`       | `credit_payments`      |
| `stg_recurly__invoice_coupon_redemption_history.sql` | ❌ it doesn't exist          |
| `stg_recurly__invoice_history.sql`              | `invoices`             |
| `stg_recurly__invoice_subscription_history.sql` | ❌ it doesn't exist       |
| `stg_recurly__line_item_history.sql`            | `line_items`           |
| `stg_recurly__plan_currency_history.sql`        | ❌ it doesn't exist                |
| `stg_recurly__plan_history.sql`                 | `plans`                |
| `stg_recurly__subscription_add_on_history.sql`  | ❌ it doesn't exist             |
| `stg_recurly__subscription_change_history.sql`  | ❌ it doesn't exist        |
| `stg_recurly__subscription_history.sql`         | `subscriptions`        |
| `stg_recurly__transaction.sql`                  | `transactions`         |
| `stg_recurly__transaction_subscription.sql`     | ❌ it doesn't exist        |

**Comment by btkcodedev** [comment]
Waiting for the review @marcosmarxm with some comments :smiley_cat: 
image

**Comment by btkcodedev** [comment]
New models possible
Fivetran - coupon_redemption_history from Airbyte - account_coupon_redemptions

**Comment by btkcodedev** [comment]
Currently, one transformation for account_balance from account has been done
New transformation possible, coupon_redemption_history from Airbyte - account_coupon_redemptions

Awaiting comment to verify, If I am going with the right path

**Comment by btkcodedev** [comment]
@marcosmarxm, Please see the last commit, I've tried to add the account_coupon_redemptions source yaml
and do the transformation, but it shows it can't be found.
Is there anything I miss, or should I remove it,
Two possible transformations are done :heavy_check_mark:

**Comment by btkcodedev** [comment]
Helloo,
I have a nested object for account id in schema account_coupon_redemptions how to add it to the source as columns?

stream: account_coupons_redemptions:
...
coupon:{
   id: "123"
},
account:{
   id: "345",
},
redemption_id: "12"


I've written directly like this, but not working | inside `source_recurly/source/account_coupon_redemptions.yaml`
image

**Comment by btkcodedev** [comment]
Update: account_notes, billing_info, line_items, credit_payments, invoices left

**Comment by btkcodedev** [comment]
Hi, @marcosmarxm, One doubt
Why is this error appearing as `not found`
Are my identifiers correct?
image

**Comment by marcosmarxm** [comment]
@btkcodedev try to get it runnig with postgres for snowflake/bigquery it is a small config problem.

**Comment by btkcodedev** [comment]
Current Update: @marcosmarxm 

URL: https://fivetran.github.io/dbt_recurly/#!/model/model.recurly_source

| **Fivetran**                                    | **Airbyte**            |
|-------------------------------------------------|------------------------|
| `stg_recurly__account_balance_history.sql`      | :heavy_check_mark:  Trasformation from `accounts`  |
| `stg_recurly__account_history.sql`              | :heavy_check_mark:  `accounts`  |
| `stg_recurly__account_note_history.sql`         | :heavy_check_mark:  `account_notes`        |
| `stg_recurly__billing_info_history.sql`         | :heavy_check_mark:  `billing_info`         |
| `stg_recurly__coupon_discount.sql`              | NULL table             |
| `stg_recurly__coupon_redemption_history.sql`    | :heavy_check_mark:  Transformation from `account_coupon_redemptions`       |
| `stg_recurly__credit_payment_history.sql`       | :heavy_check_mark:  `credit_payments`      |
| `stg_recurly__invoice_coupon_redemption_history.sql` | NULL table          |
| `stg_recurly__invoice_history.sql`              | NULL table            |
| `stg_recurly__invoice_subscription_history.sql` | NULL table    |
| `stg_recurly__line_item_history.sql`            | :heavy_check_mark:  `line_items`           |
| `stg_recurly__plan_currency_history.sql`        | NULL table            |
| `stg_recurly__plan_history.sql`                 | :heavy_check_mark:  `plans`                |
| `stg_recurly__subscription_add_on_history.sql`  | NULL table      |
| `stg_recurly__subscription_change_history.sql`  | :heavy_check_mark:  `subscriptions`      |
| `stg_recurly__subscription_history.sql`         | NULL table       |
| `stg_recurly__transaction.sql`                  |:heavy_check_mark:  `transactions`         |
| `stg_recurly__transaction_subscription.sql`     | NULL table     |

**Comment by marcosmarxm** [comment]
What is the reasoning behind creating NULL tables, or is this something you're still updating?

**Comment by btkcodedev** [comment]
@marcosmarxm, it would show error in Postgress as tables are not found, also there are no options for disabling it isn't it

Please correct if I am wrong

File: connectors/source_recurly/models/fivetran_converter/credit_payment_history.sql [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            cast(updated_at AS {{ dbt.type_timestamp() }}) AS updated_at,
+            NULL AS account_id,
+            NULL AS action,
+            cast(amount AS {{ dbt.type_float() }}) AS amount,
+            NULL AS applied_to_invoice_id,
+            cast(created_at AS {{ dbt.type_timestamp() }}) AS created_at,
+            currency,
+            NULL AS refund_transaction_id,
+            NULL AS original_credit_payment_id,
+            NULL AS original_invoice_id,
```
[the remark this answers is not in the document]
Comment (marcosmarxm) [reply connectors/source_recurly/models/fivetran_converter/credit_payment_history.sql:15]: Please check https://recurly.com/developers/api/v2021-02-25/index.html#operation/list_account_credit_payments

File: connectors/source_recurly/models/fivetran_converter/coupon_redemption_history.sql [hunk 2]
```diff
@@ -0,0 +1,61 @@
+{% if target.type == "postgres" %}
+
+    WITH tmp AS (
+        SELECT
+            id AS coupon_redemption_id,
+            cast(updated_at AS {{ dbt.type_timestamp() }}) AS updated_at,
+            NULL AS account_id,
+            NULL AS coupon_id,
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply connectors/source_recurly/models/fivetran_converter/coupon_redemption_history.sql]: Yes, absolutely, Done :white_check_mark:

File: connectors/source_recurly/models/fivetran_converter/credit_payment_history.sql [hunk 3]
```diff
@@ -0,0 +1,73 @@
+{% if target.type == "postgres" %}
+
+    WITH tmp AS (
+        SELECT
+            id AS credit_payment_id,
+            cast(updated_at AS {{ dbt.type_timestamp() }}) AS updated_at,
+            NULL AS account_id,
+            NULL AS action,
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply connectors/source_recurly/models/fivetran_converter/credit_payment_history.sql]: Done :white_check_mark:

File: connectors/source_recurly/models/fivetran_converter/credit_payment_history.sql [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            cast(amount AS {{ dbt.type_float() }}) AS amount,
+            NULL AS applied_to_invoice_id,
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply connectors/source_recurly/models/fivetran_converter/credit_payment_history.sql]: Done :white_check_mark:

File: connectors/source_recurly/models/fivetran_converter/invoice_subscription_history.sql [hunk 5]
```diff
@@ -0,0 +1,44 @@
+
+{% if target.type == "postgres" %}
+
+    WITH tmp AS (
+        SELECT
+            NULL AS invoice_id,
+            NULL AS invoice_updated_at,
+            NULL AS subscription_id,
+            NULL AS is_most_recent_record
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply connectors/source_recurly/models/fivetran_converter/invoice_subscription_history.sql]: Done ✅

File: connectors/source_recurly/models/fivetran_converter/invoice_history.sql [hunk 6]
```diff
@@ -0,0 +1,110 @@
+
+{% if target.type == "postgres" %}
+
+    WITH tmp AS (
+        SELECT
+            NULL AS invoice_id,
```
[the remark this answers is not in the document]
Comment (btkcodedev) [reply connectors/source_recurly/models/fivetran_converter/invoice_history.sql]: Done ✅

**Comment by btkcodedev** [comment]
Comments resolved, Please check @marcosmarxm :bowing_man:

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
@btkcodedev please check my last commit I made two examples to you (`account_balance_history` and `account_history`). Please feel free to contact me if you have any further question.

**Comment by btkcodedev** [comment]
Oh yes, Thanks, I'll try with others the same

**Comment by btkcodedev** [comment]
@marcosmarxm, After 5 hours of frying my brain, I could see SQL errors clearing slowly :partying_face: 
- Used your example to fill other models and Added all possible values in place of null :white_check_mark: 
- All the Snowflake and Bigquery SQL queries are cleared, only those that shows are tables doesn't exist, Please comment on how to clear those
- In postgres, there are only two errors, but the table has those fields, Please comment on this too
image


Please continue to review :bowing_man:

**Comment by btkcodedev** [comment]
In postgres error, user object is there and it has those fields
https://github.com/airbytehq/airbyte/blob/efd76b79d05bf96e5e66c72169c954e4639b4660/airbyte-integrations/connectors/source-recurly/source_recurly/manifest.yaml#L1126
Need comment on that

**Review by marcosmarxm — APPROVED** [review]
LGTM @btkcodedev there is a small change in CI to make run with bigquery and snowflake but the models looks awesome! Thanks for the contribution

**Comment by btkcodedev** [comment]
:bow Please continue to merge if it's ready.

**Pull request merged**